### DISCUSSION TITLE
How to access common constants in a best practice way? (pageContextInit not available in prerender)

### DISCUSSION THREAD

**Discussion by schaschko** [vikejs/vike#956] [Discussion]
Reproduce:

`yarn create vite-plugin-ssr`

Add in `vite.config.ts`:
`plugins: [react(), ssr({ prerender: true })]`

In `server/index.ts`:
```
@@ -29,7 +29,8 @@ async function startServer() {

   app.get('*', async (req, res, next) => {
     const pageContextInit = {
-      urlOriginal: req.originalUrl
+      urlOriginal: req.originalUrl,
+      hellotest: "Hello",
     }
     const pageContext = await renderPage(pageContextInit)
     const { httpResponse } = pageContext
```

`renderer/PageShell.tsx`
```tsx
@@ -8,6 +8,7 @@ import { Link } from './Link'
 export { PageShell }

 function PageShell({ children, pageContext }: { children: React.ReactNode; pageContext: PageContext }) {
+  console.log(pageContext.hellotest)
   return (
     <React.StrictMode>
       <PageContextProvider pageContext={pageContext}>
```

then `yarn prod`

`pageContext.hellotest` is `undefined` in prerendering scenario, with `yarn dev` not a problem (I know prerender is not called in `dev` at all, but at first I assumed it should be the same in `prod`/`dev`).

I worked around by providing `hellotest` via `onBeforeRender` for every page instead of writing middleware. The problem is, I'd like to have the same "minimal set" of constants on every page (in this example `hellotest`, which is normally fetched from db on server). So the contents of `onBeforeRender` functions for the available `*.page.server.tsx`s get redundant, as `hellotest` is fetched in everyone of them.

How to provide a minimal set of "shared constants" for the whole prerendered app in a best practice way?

**Comment by brillout** [comment]
Does https://github.com/brillout/vite-plugin-ssr/issues/962 address your issue?

Let me know whether the workaround works for you.

> The problem is, I'd like to have the same "minimal set" of constants on every page (in this example `hellotest`, which is normally fetched from db on server).

What kind of data do you want to make available to all pages?
